kaiax/builder/impl/tx_pool.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2025 The Kaia Authors
+// This file is part of the Kaia library.
+//
+// The Kaia library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Kaia library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Kaia library. If not, see <http://www.gnu.org/licenses/>.
+
+package impl
+
+import (
+	"errors"
+	"time"
+
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/kaiax"
+)
+
+var _ kaiax.TxPoolModule = (*BuilderModule)(nil)
+
+func (b *BuilderModule) PreAddTx(tx *types.Transaction, local bool) error {
+	txAndTime, ok := b.pendingBundles[tx.Hash()]
+	if ok && time.Since(txAndTime.time) < BundleLockPeriod {
+		return errors.New("Unable to add known bundle tx into tx pool during lock period")
+	}
+	return nil
+}
+
+func (b *BuilderModule) IsModuleTx(tx *types.Transaction) bool {
+	_, ok := b.pendingBundles[tx.Hash()]
+	return ok
+}
+
+func (b *BuilderModule) GetCheckBalance() func(tx *types.Transaction) error {
+	return nil
+}
+
+func (b *BuilderModule) IsReady(txs map[uint64]*types.Transaction, next uint64, ready types.Transactions) bool {
+	return true
+}
+
+func (b *BuilderModule) PreReset(oldHead, newHead *types.Header) {
+	pending, err := b.TxPool.UnlockedPending()
+	if err != nil {
+		logger.Error("Failed to get pending transactions", "err", err)
+		return
+	}
+
+	signer := types.LatestSignerForChainID(b.Backend.ChainConfig().ChainID)
+	baseFee := b.Backend.CurrentBlock().Header().BaseFee
+	txs := types.NewTransactionsByPriceAndNonce(signer, pending, baseFee)
+	arrayTxs := Arrayify(txs)
+	_, bundles := ExtractBundlesAndIncorporate(arrayTxs, b.Modules)
+
+	for _, bundle := range bundles {
+		for _, txOrGen := range bundle.BundleTxs {
+			if txOrGen.IsConcreteTx() {
+				tx, err := txOrGen.GetTx(0)
+				if err != nil {
+					logger.Error("Failed to get tx from bundle", "err", err)
+					continue
+				}
+				newTxAndTime := txAndTime{
+					tx:   tx,
+					time: time.Now(),
+				}
+				if txAndTime, ok := b.pendingBundles[tx.Hash()]; ok {
+					newTxAndTime.time = txAndTime.time
+				}
+				b.pendingBundles[tx.Hash()] = newTxAndTime
+			}
+		}
+	}
```

Comment:
Can't we simply check if tx is a module tx?
Better, can't we move this logic to add tx to `PreAddTx`?

Reply:
> Can't we simply check if tx is a module tx?

I'll try it.

> Better, can't we move this logic to add tx to PreAddTx?

This is impossible. We can judge that if a tx is in a bundle only after the tx is promoted into pending.

Reply:
Strictly speaking, yes, a tx should be promoted to be bundled.
However, a queued tx that is recognized by `txBundlingModule.IsModuleTx(tx)` is also a candidate of bundling.
Do you think it'll be problematic if we broaden the `knownTx` scope to queued bundleTx?

Reply:
Strictly you're right, but would it cause a trouble if this is applied to queue txs as well for the sake of implementation simplicity?

Reply:
There may be no problem in the event of the gasless module. But I don't know about other modules. If bundled tx pair are not promoted in the same time, it may not be enough to check only queue. Moreover bundled tx pair is not always deterministically selected. It may depend on the pending list.

Or we can add `MayBeBundled(tx) bool` to the builder module for simplicity. We cannot know if a tx will be always bundled but we can know if a tx may be bundled. It might be a good idea for all such transaction to be time out target.

Reply:
Right, I was thinking of the need for `MayBeBundled(tx) bool` as well.
For gasless module, will the condition be the same as `IsModuleTx(tx)`?

Reply:
Yes

Reply:
> Given this situation, maybe it can be difficult to implement MayBeBundled(tx) bool.
Maybe we need IsBundleTx(pending, tx) bool.

Thinking about it again, the result of `IsBundleTx(pending, tx) bool` may be different for each node because they have different pending list. Therefore, we can use only tx and pre state for the same result. 

I think it's possible to implement `MayBeBundled(tx) bool`. Even if a transaction is potential to be bundled but wouldn't be finally bundled, `MayBeBundled(tx) bool` should be true.

Reply:
I suggest `IsBundleTx(tx) bool` for the name, because the implementation will be deterministic.
For currently expected modules (gasless & auction), the result will be deterministic.